libopae/pluginmgr.c

```diff
@@ -190,15 +240,166 @@ int opae_plugin_mgr_finalize_all(void)
+		return 1;
+	}
+
+	stringified = json_object_to_json_string_ext(j_plugin_cfg, JSON_C_TO_STRING_PLAIN);
+	if (!stringified) {
+		OPAE_ERR("error getting plugin configuration");
+		return 1;
+	}
+
+	cfg->cfg_size = strlen(stringified);
+	cfg->cfg = malloc(cfg->cfg_size);
+	if (!cfg->cfg) {
+		OPAE_ERR("error allocating memory for plugin configuration");
+		cfg->cfg_size = 0;
+		return 1;
+	}
+
+	if (strncpy_s(cfg->cfg, MAX_PLUGIN_CFG_SIZE, stringified, cfg->cfg_size)) {
+		OPAE_ERR("error copying plugin configuration");
+		goto out_err;
+	}
+
+	if (strcpy_s(cfg->name, PLUGIN_NAME_MAX, name)) {
+		OPAE_ERR("error copying plugin name");
+		goto out_err;
+	}
+
+	if (strcpy_s(cfg->module, PLUGIN_NAME_MAX, json_object_get_string(j_module))) {
+		OPAE_ERR("error copying module name");
+		goto out_err;
+	}
+
+	cfg->enabled = json_object_get_boolean(j_enabled);
+	plugin_count++;
+	return 0;
+out_err:
+	if (cfg->cfg) {
+		free(cfg->cfg);
+		cfg->cfg = NULL;
+	}
+	cfg->cfg_size = 0;
+	return 1;
+}
+
+STATIC int process_cfg_buffer(const char *buffer, const char *filename)
+{
+	int num_plugins = 0;
+	int num_errors = 0;
+	int i = 0;
+	int res = 1;
+	json_object *root = NULL;
+	json_object *j_plugins = NULL;
+	json_object *j_configs = NULL;
+	json_object *j_plugin = NULL;
+	json_object *j_config = NULL;
+	const char *plugin_name = NULL;
+	enum json_tokener_error j_err = json_tokener_success;
+
+	root = json_tokener_parse_verbose(buffer, &j_err);
+	if (!root) {
+		OPAE_ERR("Error parsing config file: '%s' - %s", filename,
+			 json_tokener_error_desc(j_err));
+		goto out_free;
+	}
+
+	if (!json_object_object_get_ex(root, "plugins", &j_plugins)) {
+		OPAE_ERR("Error parsing config file: '%s' - missing 'plugins'", filename);
+		goto out_free;
+	}
+	if (!json_object_object_get_ex(root, "configurations", &j_configs)) {
+		OPAE_ERR("Error parsing config file: '%s' - missing 'configs'", filename);
+		goto out_free;
+	}
+
+	if (!json_object_is_type(j_plugins, json_type_array)) {
+		OPAE_ERR("'plugins' JSON object not array type");
+		goto out_free;
+	}
+
+	num_plugins = json_object_array_length(j_plugins);
+	num_errors = 0;
+	for (i = 0; i < num_plugins; ++i) {
+		j_plugin = json_object_array_get_idx(j_plugins, i);
+		plugin_name = json_object_get_string(j_plugin);
+
+		if (json_object_object_get_ex(j_configs, plugin_name, &j_config)) {
+			num_errors += process_plugin(plugin_name, j_config);
+		} else {
+			OPAE_ERR("Could not find plugin configuration for '%s'", plugin_name);
+			num_errors += 1;
+		}
+	}
+	res = num_errors;
+
+
+out_free:
+	json_object_put(root);
+	return res;
+
+}
+
+#define MAX_CFG_SIZE 4096
+STATIC int opae_plugin_mgr_parse_config(const char *filename)
 {
+	char buffer[MAX_CFG_SIZE] = { 0 };
+	char *ptr = &buffer[0];
+	size_t bytes_read = 0, total_read = 0;
+	FILE *fp = fopen(filename, "r");
+	if (!fp) {
+		OPAE_ERR("Error opening config file: %s", filename);
+		return 1;
+	}
+
+	while ((bytes_read = fread(ptr + total_read, 1, 1, fp))
```

Comment:
The style checker won't like this.  (It got me for this one earlier today. :))

Reply:
I _think_ this is OK. I think it depends on the usage (if vs while).

Reply:
👍